### DISCUSSION TITLE
MSC to request the host to refresh the content

### DISCUSSION THREAD

**Discussion by FabienLep** [hathach/tinyusb#584] [Q&A]
Hi,
I have an stm32 setup as a Mass storage that I can access via a computer. So far so good.
In my application, the content of the mass storage will change, and I would like the host (computer in my case) to refresh its content as soon as I have a change.
Is it possible to send a message to the host (the computer) to refresh / sync the content of the mass storage?
Thanks a lot for your help!

**Comment by cr1901** [comment]
Commenting b/c this functionality may be relevant in the _opposite_ direction. At least on MSP430, whose Mass Storage is read-only, writes to the USB drive are honored when I make changes (they shouldn't be).

**Comment by hathach** [comment]
I wonder this myself. Maybe there is a walkaround  like response a error sense code, then self-recover, that maybe trigger the host to re-scan the device. But I don't know & haven't tried it. Hopefully someone with better knowledge of SCSI command could help here. https://www.seagate.com/files/staticfiles/support/docs/manual/Interface%20manuals/100293068j.pdf .

**Reply by FabienLep** [reply]
Thank you for your answer!
The more reading I do the less I think it's possible... The issue with this workaround is that we have to wait for the host to send a message to be able to respond with an error code. So we loose the control on when the host will re-scan the device.
I'll leave it open in case someone with more SCSI knowledge has some input.

**Reply by hathach** [reply]
Host always send test unit ready before any other op. You can try to return failed with that. But scsi commands are very comprehensive with tons of variants, you could try to poke at linux code to see when it does a rescan to get some idea for solution

**Comment by hathach** [comment]
Update: I think test unit ready can be used as walkaround to force host to update. You could simulate and ejected event by return false to test unit ready and set sense key to NOT READY (subkey not found). Then report as ready again, it would force host to refresh

```C
// Invoked when received Test Unit Ready command.
// return true allowing host to read/write this LUN e.g SD card inserted
bool tud_msc_test_unit_ready_cb(uint8_t lun)
{
  (void) lun;

  // RAM disk is ready until ejected
  if (ejected) {
    tud_msc_set_sense(lun, SCSI_SENSE_NOT_READY, 0x3a, 0x00);
    return false;
  }

  return true;
}
```
